### PR TITLE
LSP 3.16 updates

### LINKED ISSUE #170: Progress report capabilities
Hi there,

I am trying to setup a progress report from the LS to Visual Code.

I implemented the following functions :

```python
@diplomat_server.command(DiplomatLanguageServer.CMD_TST_PROGRESS_STRT)
def on_start_progress(ls : DiplomatLanguageServer, *args):
	initiate_progress(ls)
	print("Start progress on uuid",ls.progress_uuid)
	ls.send_notification("$/progress",WorkDoneProgressParams(token=ls.progress_uuid,value=WorkDoneProgressBegin(title = "Working hard")))


@diplomat_server.command(DiplomatLanguageServer.CMD_TST_PROGRESS_STOP)
def on_stop_progress(ls : DiplomatLanguageServer, *args):
	print("Stop progress")

	ls.send_notification("$/progress",WorkDoneProgressParams(token=ls.progress_uuid, value=WorkDoneProgressEnd(message = "Stuff done.")))

@diplomat_server.command(WINDOW_WORK_DONE_PROGRESS_CREATE)
def initiate_progress(ls : DiplomatLanguageServer, *args) -> ProgressToken:
	ls.progress_uuid = uuid.uuid4().int
	return ls.progress_uuid
```

When I execute `on_start_progress` and `on_stop_progress` from a client command, I do get the appropriates prints on the python side, however, I have no reaction at all even though I expect to have an indefinite progress bar.

I haven't found any resource about this in the documentation so is there a limitation that I'm not aware of or an obvious mistake 
? I suspect my `initiate_progress` command or its call are the problem, but I have no clue about how to do otherwise...

If there is a place to  ask for help which is more appropriate than GH issues, please let me know. 

Best regards,
Julien Faucher

### PR DESCRIPTION
**Pull request by danixeee**
## Description (e.g. "Related to ...", etc.)

- [x] Progress bar support
- [x] Trace log support
- [X] LSP 3.16 model updates
- [x] Tests

Closes #170, #162, #115.

## Code review checklist (for code reviewer to complete)

- [ ] Pull request represents a single change (i.e. not fixing disparate/unrelated things in a single PR)
- [ ] Title summarizes what is changing
- [ ] Commit messages are meaningful (see [this][commit messages] for details)
- [ ] Tests have been included and/or updated, as appropriate
- [ ] Docstrings have been included and/or updated, as appropriate
- [ ] Standalone docs have been updated accordingly
- [ ] [CONTRIBUTORS.md][contributors] was updated, as appropriate
- [ ] Changelog has been updated, as needed (see [CHANGELOG.md][changelog])

[changelog]: https://github.com/openlawlibrary/pygls/blob/master/CHANGELOG.md
[commit messages]: https://chris.beams.io/posts/git-commit/
[contributors]: https://github.com/openlawlibrary/pygls/blob/master/CONTRIBUTORS.md

### CODE REVIEW

**Comment by danixeee** [comment]
To address, from specs:

> Notification and requests whose methods start with ‘$/’ are messages which are protocol implementation dependent and might not be
implementable in all clients or servers. For example if the server implementation uses a single threaded
synchronous programming language then there is little a server can do to react to a $/cancelRequest notification. 
If a server or client receives notifications starting with ‘$/’ it is free to ignore the notification.
If a server or client receives a requests starting with ‘$/’ it must error the request with error code MethodNotFound (e.g. -32601).

> A request that got canceled still needs to return from the server and send a response back. It can not be left open / hanging.
This is in line with the JSON RPC protocol that requires that every request sends a response back.
In addition it allows for returning partial results on cancel.
If the request returns an error response on cancellation it is advised to set the error code to ErrorCodes.RequestCancelled

File: pygls/protocol.py [hunk 1]
```diff
@@ -683,22 +679,34 @@ def get_configuration(self, params, callback):
...
-        return asyncio.wrap_future(self.get_configuration(params, None))
+        return asyncio.wrap_future(self.get_configuration(params))
+
+    def log_trace(self, message: str, verbose: Optional[str] = None) -> None:
+        """Sends trace notification to the client."""
+        if self.trace == Trace.Off:
+            return
+
+        params = LogTraceParams(message=message)
+        if self.trace == Trace.Verbose and verbose:
```
Comment (renatav) [inline pygls/protocol.py:698]: Very minor thing - to me it's a lot more intuitive to check `verbose` first and then if equality

**Review by renatav — COMMENTED** [review]
Minor comments, looks good to me

File: pygls/lsp/types/language_features/document_symbol.py [hunk 2]
```diff
@@ -61,18 +61,32 @@ class SymbolKind(enum.IntEnum):
     TypeParameter = 26
 
 
+class SymbolTag(enum.IntEnum):
+    Deprecated = 1
```
[the remark this answers is not in the document]
Comment (danixeee) [reply pygls/lsp/types/language_features/document_symbol.py:65]: Yup, that's from the [specs](https://microsoft.github.io/language-server-protocol/specifications/specification-3-17/#symbolTag).

File: pygls/progress.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+        self.tokens: Dict[ProgressToken, None] = {}
+
+    def create(self, token: ProgressToken, callback=None) -> Future:
+        if token in self.tokens:
+            raise Exception("Token is already registered!")  # TODO
+
+        def on_created(*args, **kwargs):
+            self.tokens[token] = None
+            if callback:
```
[the remark this answers is not in the document]
Comment (danixeee) [reply pygls/progress.py:26]: Not sure, I can make it explicit.

**Review by renatav — APPROVED** [review]

**Pull request merged**